rocks-lib/src/lockfile/mod.rs

```diff
@@ -115,40 +132,162 @@ impl LocalPackage {
         self.pinned
     }
 
-    pub fn dependencies(&self) -> Vec<&String> {
+    pub fn dependencies(&self) -> Vec<&LocalPackageId> {
         self.dependencies.iter().collect()
     }
 
-    pub fn constraint(&self) -> LockConstraint {
-        match &self.constraint {
-            // Safe to unwrap as the data can only end up in the struct as a valid constraint
-            Some(constraint) => LockConstraint::Constrained(
-                constraint
-                    .parse()
-                    .expect("invalid constraint in LuaPackage"),
+    pub fn to_package(&self) -> RemotePackage {
+        RemotePackage::new(self.name.clone(), self.version.clone())
+    }
+
+    pub fn into_package_req(self) -> PackageReq {
+        RemotePackage::new(self.name, self.version).into_package_req()
+    }
+}
+
+// TODO(vhyrro): Move to `package/local.rs`
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
+pub struct LocalPackage {
+    pub spec: LocalPackageSpec,
+    pub hashes: LocalPackageHashes,
+}
+
+#[cfg_attr(feature = "lua", derive(FromLua,))]
+#[derive(Debug, Serialize, Deserialize, Clone)]
+struct LocalPackageIntermediate {
+    name: PackageName,
+    version: PackageVersion,
+    pinned: PinnedState,
+    dependencies: Vec<LocalPackageId>,
+    constraint: Option<String>,
+    hashes: LocalPackageHashes,
+}
```

Comment:
question: could we make do without this intermediary struct? I think it makes the deserialization flow quite hard to follow - is there anything blocking us from doing the custom deserialization magic directly for LocalPackage? I can't pick it up just from looking at the code :smile:

Reply:
Ah I see, this is about `FromLua`. Maybe we should just implement FromLua ourselves here to solve this issue? It's a shame the derive macro doesn't allow cheeky things like `flatten`/`skip_serializing` like you can in serde hah.

Reply:
> Maybe we should just implement FromLua ourselves here to solve this issue?

That's what I'm doing (also for `Serialize` and `Deserialize`).
The nice thing about this approach is that those 3 trait implementations can reuse the `TryFrom` and `From` implementations.

```rust
impl<'de> Deserialize<'de> for LocalPackage {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: serde::Deserializer<'de>,
    {
        LocalPackage::try_from(LocalPackageIntermediate::deserialize(deserializer)?)
            .map_err(de::Error::custom)
    }
}

impl Serialize for LocalPackage {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: serde::Serializer,
    {
        LocalPackageIntermediate::from(self).serialize(serializer)
    }
}

#[cfg(feature = "lua")]
impl FromLua for LocalPackage {
    fn from_lua(value: mlua::Value, lua: &mlua::Lua) -> mlua::Result<Self> {
        LocalPackage::try_from(LocalPackageIntermediate::from_lua(value, lua)?)
            .map_err(|err| mlua::Error::DeserializeError(format!("{}", err)))
    }
}
```

Reply:
let's make LocalPackageSpec `pub(crate)`, as you intended :)

Reply:
Done